src/routes/slack.js

```diff
@@ -0,0 +1,183 @@
+      const songs = songSearch.getResults();
+      result = songs.then(resultSongs => SongTemplate.getSongCardArray(resultSongs, username));
+      break;
+    case 'artist':
+      const artistSearch = new ArtistSearch(keyword, 5);
+      artistSearch.search();
+      const artists = artistSearch.getResults();
+      result = artists.then(resultArtists => ArtistTemplate.getArtistCardArray(resultArtists, username));
+      break;
+    case 'album':
+      const albumSearch = new AlbumSearch(keyword, 5);
+      albumSearch.search();
+      const albums = albumSearch.getResults();
+      result = albums.then(resultAlbums => AlbumTemplate.getAlbumCardArray(resultAlbums, username));
+      break;
+    default:
+      result = new Promise((resolve, reject) => reject());
+  }
+
+  result
+    .then((cards: CardTemplate[]) => {
+      const responseText: string = (cards.length > 0) ?
+        'Okay, this is what I found:' :
+        'I\'m sorry, I didn\'t find anything. ☹️';
+
+      const response: ResponseTemplate = new ResponseTemplate(responseText, cards);
+      res.status(200).send(response);
+      return;
+    })
+    .catch(reason => {
+      logger.error(reason);
+      res.status(200).send('Sorry, something went terribly wrong. ☹️');
+      return;
+    });
+
+  logger.info({
+    action: 'search',
+    platform: 'Slack',
+    command: param,
+    term: keyword
+  });
+
+});
+
+/**
+ * Main router endpoint.
+ * Makes the call to the controller and returns the message
+ */
+router.post('/postback', (req, res) => {
+  res.sendStatus(200);
+
+  const { body } = req;
+  const payload: any = JSON.parse(body.payload);
+  const value: any = JSON.parse(payload.actions[0].value);
+  const attachments: ResponseTemplate[] = value.attachments;
+  const text: string = value.text;
+  const attachment: string = JSON.stringify(attachments);
+
+  SlackCredentialModel.retrieve(payload.team.id)
+    .then(credential => {
+      rp({
+        method: 'GET',
+        uri: 'https://slack.com/api/chat.postMessage',
+        qs: {
+          token: credential.botAccessToken,
+          channel: payload.channel.id,
+          text: text,
+          as_user: false,
+          link_names: true,
+          attachments: attachment
+        }
+      })
+      .catch(reason => logger.error(reason));
+    });
+
+  logger.info({
+    action: 'post',
+    Platform: 'Slack',
+    attachment: attachments
+  });
+
+});
+
+router.get('/auth', (req, res) => {
+  const state: string = req.query.state;
+  const code: string = req.query.code;
+
+  rp({
+    method: 'GET',
+    uri: 'https://slack.com/api/oauth.access',
+    qs: {
+      client_id: process.env.SLACK_CLIENT_ID,
+      client_secret: process.env.SLACK_CLIENT_SECRET,
+      code,
+      redirect_uri: process.env.SLACK_REDIRECT_URI
+    }
+  })
+    .then(response => {
+      const credentialResponse = JSON.parse(response);
+      if (credentialResponse.ok) {
+        const credential = new SlackCredential();
```

Comment:
I know it will be kind of long, but you may be able to pass the credentials stuff to the constructor, so you avoid the temporal coupling and you make the code looks cleaner.
```js
const credentials = new SlackCredentials({
  accessToken: credentialResponse.access_token,
  teamName: credentialResponse.team_name,
  teamId: credentialResponse.team_id,
  botUserId: credentialResponse.bot.bot_user_id,
  botAccessToken: credentialResponse.bot.bot_access_token
});

SlackCredentialModel.save(credential);
```